src/Codeception/Util/JsonType.php

```diff
@@ -126,14 +126,25 @@ protected function typeComparison($data, $jsonType)
                 }
                 continue;
             }
-            $matchTypes = preg_split("#(?![^]\(]*\))\|#", $type);
-            $matched = false;
+
+            $regexes = [];
+
+            preg_match_all('/:regex\((.*?[!"#$%&\'*+,.\/:;=?@^_`|~-])\)/', $type, $regexes);
+
+            $type = preg_replace_callback('/:regex\((.*?[!"#$%&\'*+,.\/:;=?@^_`|~-])\)/', function ($m) {
```

Comment:
could you explain a bit of this magic?

Reply:
So what's happening here's it's matching for everything up to the regex closing delimiter (This should cover 99.999% of the delimiter usage - There are some strange ones like '<'...'>' and '{'...'}' which I've only come across a couple times in my lifetime.). What was happening before was that stuff like ':' was causing the match to match prematurely, so adding the closing delimiter match after wildcard and before the ')' character, it shouldn't get tripped up.

Reply:
Just use the first character as a delimiter, that's what preg_match does.

Reply:
Well the offending code was preg_split, which was looking for ':', which would break if ':' was in the regex filter definition. So this is what I first thought of. There is the (?R) recursive pattern, but I would have to look over that again, since it's probably been over 10 years since I've used that.

Reply:
@DavertMik @Naktibalda I've updated my explanation on the how and why.